### PR TITLE
Stable sort via bit magic

### PR DESCRIPTION
**Pull request by kentslaney**
I rolled back the `match_sort` implementation because of the `AndOp`/`OrOp` thing. Not quite sure where that kind of `SelectOp` is used?

### CODE REVIEW

File: stablehlo_coreml/converter.py [hunk 1]
```diff
@@ -418,21 +416,27 @@ def op_sort(self, context: TranslationContext, op: SortOp):
         if sort_keys is None:
             raise ValueError("Unrecognized comparator format")
 
+        def precast_gather(tensor, indices):
+            if tensor.dtype == types.int32:
```
Comment (kasper0406) [inline stablehlo_coreml/converter.py]: Fine workaround. Fysa, I think this is caused by a bug in coremltools that should be fixed by: https://github.com/apple/coremltools/pull/2626

File: stablehlo_coreml/sort_utils.py [hunk 2]
```diff
@@ -1,212 +1,227 @@
...
+    ieee754 = {np.float32: (8, 23), np.float16: (5, 10)}
+    assert types.is_float(x.dtype) and width in ieee754
+    exponent, fraction = ieee754[width]
+    e_bias = 2 ** (exponent - 1) - 1
+    e_offset = 2 ** fraction
+
+    # Detect special values
+    # mb.is_nan/is_inf are not available, use comparisons
+    is_nan = mb.not_equal(x=x, y=x)
+    is_inf = mb.equal(x=mb.abs(x=x), y=width(float('inf')))
```
Comment (kasper0406) [inline stablehlo_coreml/sort_utils.py]: nit: I think you can use `np.inf` instead of `float('inf')` which looks slightly cleaner

**Review by kasper0406 — COMMENTED** [review]
Good work getting this to work - some very impressive and complicated code. I won't really pretend to understand how it works, and I've added some comments where my understanding struggles the most.

File: stablehlo_coreml/converter.py (hunk 1 above)
Comment (kentslaney) [reply stablehlo_coreml/converter.py]: Oh, I actually rely on that bug for gathering the full 16 bits in stable sort (the part where I add `0x1_0000` to negative numbers). I would be able to transfer the same amount of info if they moved the sign bit over instead, but I've been avoiding that bit of int32s (thus the shift by `width - 1 - mask`). It's maybe possible to make it behave the same either way, but haven't thought it through yet.

I can add some of this context as comments too, I'm just used to and tired from working alone.

File: stablehlo_coreml/converter.py [hunk 3]
```diff
@@ -418,21 +416,27 @@ def op_sort(self, context: TranslationContext, op: SortOp):
...
+                tensor = mb.cast(x=tensor, dtype="int16")
+                res = mb.gather_along_axis(x=tensor, indices=indices, axis=sort_dim)
+                return mb.cast(x=res, dtype="int32")
+            else:
+                return mb.gather_along_axis(x=tensor, indices=indices, axis=sort_dim)
+
         # Apply the sort
         sort_dim, (key, ascending) = op.dimension.value, sort_keys[-1]
-        indices = mb.argsort(x=key, axis=sort_dim, ascending=ascending)
+        indices = stable_argsort(x=key, axis=sort_dim, ascending=ascending)
```
[the remark this answers is not in the document]
Comment (kentslaney) [reply stablehlo_coreml/converter.py]: Oh yeah, I missed that it's less relevant for a single key sort. I think some of your code also had a check for HLO SortOp's `is_stable`, but I'm not sure if that's just permanently true in practice.

**Comment by kentslaney** [comment]
>```python
> def test_argsort():
>     def sort_dim_0(k1):
>         k2 = jnp.arange(k1.shape[0], dtype=jnp.int32)
>         return jax.lax.sort([k1, k2], dimension=0, num_keys=2)
>
>     run_and_compare_specific_input(sort_dim_0, (
>         jnp.array([10, 5, 10, 0x1_0010, 10, 0x5_0000, 10, 5, 10, 0x1_0000, 5], dtype=jnp.int32),
>     ))
>```
> ```python
> E            ACTUAL: array([ 0,  0,  5,  5,  5, 10, 10, 10, 10, 10, 16], dtype=int32)
> E            DESIRED: array([     5,      5,      5,     10,     10,     10,     10,     10,
> E                   65536,  65552, 327680], dtype=int32)
> ```

@kasper0406 This is actually because the gather of `k1` from sorting `k2` overflows, which I don't think I can fix (besides in the trivial case where there are no changes). In the original it's `argsort` instead of `sort`, so you can tell it's stable because the elements have been moved the minimum amount possible.

File: stablehlo_coreml/sort_utils.py [hunk 4]
```diff
@@ -1,212 +1,227 @@
...
-    def get_op(val):
-        if isinstance(val.owner, ir.Block):
-            return None
-        return val.owner.opview
-
+from coremltools.converters.mil.mil import Builder as mb
+from coremltools.converters.mil.mil import types
+
+
+def bitcast_fp(x):
```
[the remark this answers is not in the document]
Comment (kasper0406) [reply stablehlo_coreml/sort_utils.py]: Right. I just have a hard time convincing myself if it is one of those things which is always doing the correct thing, or doing the correct thing "most of the time"

**Comment by kentslaney** [comment]
I can't seem to independently reproduce the overflow behavior that I thought was causing the problem

```python
import numpy as np
import coremltools as ct
from coremltools.converters.mil.mil import Builder as mb
from coremltools.converters.mil.mil import types

def check_size(n, target):
    @mb.program(input_specs=[
        mb.TensorSpec(shape=(n,), dtype=types.int32),
        mb.TensorSpec(shape=(1,), dtype=types.int32)
    ])
    def prog(x, indices):
        out = mb.gather_along_axis(x=x, indices=indices, axis=0, name="output")
        return out

    model = ct.convert(prog, source='milinternal', convert_to="mlprogram", compute_precision=ct.precision.FLOAT32)

    x_val = np.arange(n, dtype=np.int32)
    indices_val = np.array([target], dtype=np.int32)

    prediction = model.predict({"x": x_val, "indices": indices_val})
    result = prediction["output"]

    print(f"{n:#_x} {target:#_x} {result[0]:#_x}")

check_size(0x10_0010, 0x10_0008)
```

**Comment by kentslaney** [comment]
Issue description starting from the [`sort-gather-overflow` branch](https://github.com/kentslaney/stablehlo-coreml/tree/f7070c72be3a6a575158c9a2b52bd599bcc44d03)

`env/bin/python tests/mb_gather_overflow.py` suggests that `mb.gather_along_axis` shouldn't have a problem with operands that don't fit in 16 bits. Why, then, does `env/bin/pytest tests -k test_gather_cast` seem to wrap when the gathered input is larger than 16 bits?

Interestingly, switching the first `stable_argsort` to `mb.argsort` in `op_sort` removes the wrapping issue (though introduces an order issue) even though the second input with the wrapping problem only goes through the final `mb.gather_along_axis` call and the behavior persists with an explicit cast of indices to int32, which should remove any discrepancies in input types.

**Comment by kentslaney** [comment]
I've yet to check whether patching to apple/coremltools#2626 makes a difference (dumping note-to-self progress updates here because of an intermission)

**Comment by kentslaney** [comment]
I'm getting [inconsistent results](https://github.com/kentslaney/stablehlo-coreml/blob/e45274323f14b316a0b167f477223d21851b5cf1/consistency.sh#L3-L13) and sometimes a segfault on what I expect to be a deterministic pipeline.

**Comment by kentslaney** [comment]
>Btw. if you have Discord / other chat client, I'm happy to iterate there as well, if easier.

My Discord is @1i1iii1 ([relevant](https://xkcd.com/1105))
